contracts/cw3-fixed-multisig/src/contract.rs

```diff
@@ -0,0 +1,267 @@
+use cosmwasm_std::{
+    to_binary, Api, Binary, CosmosMsg, Empty, Env, Extern, HandleResponse, HumanAddr, InitResponse,
+    Querier, StdError, StdResult, Storage,
+};
+use cw2::set_contract_version;
+
+use crate::msg::{HandleMsg, InitMsg, QueryMsg};
+use crate::state::{
+    ballots, ballots_read, config, config_read, next_id, proposal, proposal_read, voter_weight,
+    voter_weight_read, Ballot, Config, Proposal,
+};
+use cw0::Expiration;
+use cw3::{ProposalResponse, Status, ThresholdResponse, Vote, VoteResponse};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:cw3-fixed-multisig";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+pub fn init<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    _env: Env,
+    msg: InitMsg,
+) -> StdResult<InitResponse> {
+    set_contract_version(&mut deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    let total_weight = msg.voters.iter().map(|v| v.weight).sum();
+    let cfg = Config {
+        required_weight: msg.required_weight,
+        total_weight,
+        max_voting_period: msg.max_voting_period,
+    };
+    config(&mut deps.storage).save(&cfg)?;
+
+    // add all voters
+    let mut voters = voter_weight(&mut deps.storage);
+    for voter in msg.voters.iter() {
+        let key = deps.api.canonical_address(&voter.addr)?;
+        voters.save(key.as_slice(), &voter.weight)?;
+    }
+    Ok(InitResponse::default())
+}
+
+pub fn handle<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    msg: HandleMsg,
+) -> StdResult<HandleResponse<Empty>> {
+    match msg {
+        HandleMsg::Propose {
+            title,
+            description,
+            msgs,
+            earliest,
+            latest,
+        } => handle_propose(deps, env, title, description, msgs, earliest, latest),
+        HandleMsg::Vote { proposal_id, vote } => handle_vote(deps, env, proposal_id, vote),
+        HandleMsg::Execute { proposal_id } => handle_execute(deps, env, proposal_id),
+        HandleMsg::Close { proposal_id } => handle_close(deps, env, proposal_id),
+    }
+}
+
+pub fn handle_propose<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    title: String,
+    description: String,
+    msgs: Vec<CosmosMsg>,
+    // we ignore earliest
+    _earliest: Option<Expiration>,
+    latest: Option<Expiration>,
+) -> StdResult<HandleResponse<Empty>> {
+    // only members of the multisig can create a proposal
+    let raw_sender = deps.api.canonical_address(&env.message.sender)?;
+    let weight = voter_weight_read(&deps.storage)
+        .may_load(raw_sender.as_slice())?
+        .ok_or_else(StdError::unauthorized)?;
+
+    // TODO: using max as default here, also enforce max
+    let cfg = config_read(&deps.storage).load()?;
+    let expires = latest.unwrap_or(cfg.max_voting_period);
```

Comment:
No need to enforce that `latest` (if it is set) is less than or equal than `cfg.max_voting_period`?

Reply:
Also, if and when `earliest` is not ignored, some consistency checks (`earliest` < `latest` <= `max_voting_period`) may be a good idea.

Reply:
Definitely, that is the meaning of the TODO - to set a max, and I abused it as default.